api/app/controllers/spree/api/shipments_controller.rb

```diff
@@ -40,13 +40,31 @@ def select_shipping_method
 
       def create
         authorize! :create, Shipment
-        quantity = params[:quantity].to_i
+
         @shipment = @order.shipments.create(stock_location_id: params.fetch(:stock_location_id))
-        @order.contents.add(variant, quantity, { shipment: @shipment })
 
-        @shipment.save!
+        if passing_deprecated_params_on_create?
+          Spree::Deprecation.warn <<~MSG
+          Passing `quantity` or `variant_id` to
 
-        respond_with(@shipment.reload, default_template: :show)
+              POST /api/shipments
+
+          is deprecated and won't be allowed anymore starting from Solidus 4.0.
+          Instead, create an empty shipment and add items to it subsequently using
+          the dedicated endpoint:
+
+              PUT /api/shipments/{shipment_number}/add
+
+          MSG
+
+          quantity = params[:quantity].to_i
+          variant = Spree::Variant.unscoped.find(params[:variant_id])
+          @order.contents.add(variant, quantity, { shipment: @shipment })
+          @shipment.save!
+          @shipment.reload
```

Comment:
Do we need `reload` here?

Reply:
I don't know (see https://github.com/solidusio/solidus/pull/4387/commits/d018c6e4f0c0b699a8b65012457332bef96627db). But being code that we are going to remove soon, I think it's not even worth spending too much time to understand if there was a reason for that reload to be there.

Reply:
Yeah, you're right :+1: